Driver.cpp

```diff
@@ -71,49 +71,130 @@ hidl_vec<Capabilities::OperandPerformance> nonExtensionOperandPerformance(
     ret1 = ret;
-    
+
     return ret1;
 }
 
+static sp<BasePreparedModel> ModelFactory(const char* name, const Model& model) {
+    sp<BasePreparedModel> driverPreparedModel = NULL;
+
+    if (strcmp(name, "CPU") == 0)
+        driverPreparedModel = new CpuPreparedModel(model);
+    else if (strcmp(name, "GNA") == 0)
+        driverPreparedModel = new GnaPreparedModel(model);
+    return driverPreparedModel;
+}
+
 // For HAL-1.0 version
 Return<void> Driver::getCapabilities(getCapabilities_cb cb) {
     ALOGV("Entering %s", __func__);
-
-    return Void();
+    return getCapabilities_1_3(
+        [&](V1_3::ErrorStatus error, const V1_3::Capabilities& capabilities) {
+            // TODO(dgross): Do we need to check compliantWithV1_0(capabilities)?
+            cb(convertToV1_0(error), convertToV1_0(capabilities));
+        });
 }
 
 Return<void> Driver::getSupportedOperations(const V1_0_Model& model, getSupportedOperations_cb cb) {
     ALOGV("Entering %s", __func__);
-
-    return Void();
+    if (!validateModel(model)) {
+        ALOGE("NNERR: %s failed at line no: %d\n", __func__, __LINE__);
+        cb(V1_0::ErrorStatus::INVALID_ARGUMENT, {});
+        return Void();
+    }
+    return getSupportedOperations_1_3(
+        convertToV1_3(model), [&](V1_3::ErrorStatus status, const hidl_vec<bool>& supported) {
+            cb(convertToV1_0(status), supported);
+        });
 }
 
 Return<ErrorStatus> Driver::prepareModel(const V1_0_Model& model,
                                          const sp<V1_0::IPreparedModelCallback>& callback) {
     ALOGV("Entering %s", __func__);
+    if (callback.get() == nullptr) {
+        ALOGE("invalid callback passed to prepareModel");
+        return ErrorStatus::INVALID_ARGUMENT;
+    }
+    if (!validateModel(model)) {
+        callback->notify(ErrorStatus::INVALID_ARGUMENT, nullptr);
+        return ErrorStatus::INVALID_ARGUMENT;
+    }
+
+    // TODO: make asynchronous later
+    sp<BasePreparedModel> driverPreparedModel =
+        ModelFactory(mDeviceName.c_str(), convertToV1_3(model));
+    if (driverPreparedModel == NULL) {
+        ALOGE("failed to create preparedmodel");
+        return ErrorStatus::INVALID_ARGUMENT;
+    }
+    for (auto opn : model.operations) dumpOperation(opn);
+
+    if (!driverPreparedModel->initialize(convertToV1_3(model))) {
+        ALOGE("failed to initialize preparedmodel");
+        callback->notify(ErrorStatus::INVALID_ARGUMENT, nullptr);
+        return ErrorStatus::NONE;
+    }
 
+    callback->notify(ErrorStatus::NONE, driverPreparedModel);
+    ALOGV("Exiting %s", __func__);
     return ErrorStatus::NONE;
 }
 
 // For HAL-1.1 version
 Return<void> Driver::getCapabilities_1_1(getCapabilities_1_1_cb cb) {
     ALOGV("Entering %s", __func__);
-
-    return Void();
+    return getCapabilities_1_3(
+        [&](V1_3::ErrorStatus error, const V1_3::Capabilities& capabilities) {
+            // TODO(dgross): Do we need to check compliantWithV1_1(capabilities)?
```

Comment:
Seems like the TODO comment was inherited from the sampleDriver. Do we need to call this out. What is the licensing rules if we are reusing the code from the SampleDriver?

Reply:
Yes, we need to call it for backward compatibility.
I am not aware about the licensing rules. It says sampleDriver ,I assumed its an example on how to implement the driver interfaces. So referred it . But will check the licensing rules once.